proposals/nnnn-mt-service-discovery.md

```diff
@@ -0,0 +1,134 @@
+
+A new class `TcpServiceAdvertiser` is added under `transport_adapter` namespace.
+
+The class includes a reference implementation to start and stop advertising. When starting advertisement, it generates a config file under /etc/avahi/service/ directory and send a SIGHUP signal to avahi-daemon. When stopping advertisement, it removes the config file and send a SIGHUP signal again.
+
+`TcpClientListener` class is extended to call the methods of `TcpServiceAdvertiser` during `StartListening()` and `StopListening()`.
+
+smartDeviceLink.ini should have a new entry to enable/disable the advertisement feature. Also, it is preferable to add another entry to configure the service name. `TcpServiceAdvertiser` uses the service name and TCP port number written in the .ini file to set up the config file under /etc/avahi/service/ directory.
+
+
+## Potential downsides
+
+* This proposal does not offer a means to select a head unit when there are multiple head units on a Wi-Fi network. Proxy simply tries to connect the first head unit that has been discovered. The author thinks it is very uncommon that more than one head units are on a single Wi-Fi network, except for development/debugging in a lab.
+* Head units have to add a service such as avahi-daemon for service discovery mechanism.
+* Service discovery mechanism consumes a little bandwidth of Wi-Fi network.
+
+
+## Impact on existing code
+
+This proposal should have little impact on existing code since most of the codes are for newly added classes.
+
+
+## Out of scope of this proposal
+
+Automatic configuration of Wi-Fi network credential (SSID and passphrase) is not included in this proposal. It is assumed that the credential is configured and stored on the phone and/or on the head unit beforehand.
+
+There are multiple types of Wi-Fi network topology; the head unit acts as an access point, the phone acts as an access point by using Wi-Fi tether feature, and the user brings a Wi-Fi hotspot device in the car. The service discovery mechanism should work in either case, so network topology discussion is out of scope of this proposal.
+
+
+## Discussions
+
+It is desirable that a SDL app automatically launches and starts service discovery once the phone is connected to a head unit over Wi-Fi. This sounds quite difficult:
+- Starting from iOS 10, an app will be launched in background when an external iAP accessory (head unit in this case) is connected over Bluetooth or USB. Or, a head unit may have custom implementation to launch specific app using iAP's launch request. Either case, iAP connection is required before launching an app.
+- Previously Android app could receive `WIFI_STATE_CHANGED_ACTION` (android.net.wifi.WIFI\_STATE\_CHANGED) broadcast intent, which is similar to ACTION\_ACL\_CONNECTED in the case of Bluetooth. Starting from Android O, apps are not able to register broadcast receivers for implicit broadcasts. ACTION\_ACL\_CONNECTED is an exception(*3) and apps can still receive it.

```

Comment:
Can you please add https://developer.android.com/guide/components/broadcast-exceptions.html as a link here instead of as a footnote in this section?  Perhaps `ACTION_ACL_CONNECTED is an exception (see [here](https://developer.android.com/guide/components/broadcast-exceptions.html)) and apps..."